File: 프로그래머스/1/142086. 가장 가까운 같은 글자/가장 가까운 같은 글자.java
class Solution {
    public int[] solution(String s) {
        int[] answer = new int[s.length()];
        int[] alpha = new int[26];
        
        for (int i = 0; i < 26; i++) {
            alpha[i] = -2;    
        }
        
        for (int i = 0; i < s.length(); i++) {
            if (alpha[s.charAt(i) - 97] == -2) {
                answer[i] = -1;
                alpha[s.charAt(i) - 97] = i;
            } else {
                answer[i] = i - alpha[s.charAt(i) - 97];
                alpha[s.charAt(i) - 97] = i;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/1/142086. 가장 가까운 같은 글자/가장 가까운 같은 글자.js
// 처음 풀이
// function solution(s) {
//     let answer = [];
//     const dict = new Map();
    
//     const arr = s.split('');
//     for (let i = 0; i < arr.length; i++) {
//         if (!dict.has(arr[i])) {
//             dict.set(arr[i], i);
//             answer.push(-1);
//         } else {
//             const location = i - dict.get(arr[i]);
//             dict.set(arr[i], i);
//             answer.push(location);
//         }
//     }
    
//     return answer;
// }

// 불필요한 연산 제거
function solution(s) {
    let answer = [];
    const dict = new Map();

    for (let i = 0; i < s.length; i++) {
        if (!dict.has(s[i])) {
            dict.set(s[i], i);
            answer.push(-1);
        } else {
            const location = i - dict.get(s[i]);
            dict.set(s[i], i);
            answer.push(location);
        }
    }
    
    return answer;
}